crates/sui-verifier/src/char_type_verifier.rs

```diff
@@ -0,0 +1,227 @@
+//! validator:
+//!
+//! - it has only one ability: drop
+//! - it has only one arbitrarily named field of type boolean (since Move structs cannot be empty)
+//! - its definition does not involve type parameters
+//! - its only instance in existence is passed as an argument to the module initializer
+//! - it is never instantiated anywhere in its defining module
+
+use move_binary_format::{
+    access::ModuleAccess,
+    binary_views::BinaryIndexedView,
+    file_format::{
+        Ability, AbilitySet, Bytecode, CompiledModule, FunctionDefinition, FunctionHandle,
+        SignatureToken, StructDefinition, StructHandle, TypeSignature,
+    },
+};
+use move_core_types::{ident_str, language_storage::ModuleId};
+use sui_types::{
+    base_types::{TX_CONTEXT_MODULE_NAME, TX_CONTEXT_STRUCT_NAME},
+    error::ExecutionError,
+    SUI_FRAMEWORK_ADDRESS,
+};
+
+use crate::{verification_failure, INIT_FN_NAME};
+
+pub fn verify_module(module: &CompiledModule) -> Result<(), ExecutionError> {
+    // In Sui's framework code there is an exception to the characteristic type rule - we have a SUI
+    // type in the sui module but it is instantiated and the module has no initializer (the reason
+    // for it is that the SUI coin is only instantiated during genesis). It is easiest to simply
+    // special-case this module particularly that this is framework code and thus deemed correct.
+    if ModuleId::new(SUI_FRAMEWORK_ADDRESS, ident_str!("sui").to_owned()) == module.self_id() {
+        return Ok(());
+    }
+
+    let view = BinaryIndexedView::Module(module);
+    let mod_handle = view.module_handle_at(module.self_module_handle_idx);
+    let mod_name = view.identifier_at(mod_handle.name).as_str();
+    let struct_defs = &module.struct_defs;
+    let mut char_type_candidate = None;
+    for def in struct_defs {
+        let struct_handle = module.struct_handle_at(def.struct_handle);
+        let struct_name = view.identifier_at(struct_handle.name).as_str();
+        if struct_name.to_ascii_uppercase() == struct_name
+            && mod_name.to_ascii_uppercase() == struct_name
+        {
+            verify_char_type(module, struct_name, struct_handle, def)
+                .map_err(verification_failure)?;
+            // if we reached this point, it means we have a legitimate characteristic type candidate
+            // and we have to make sure that both the init function's signature reflects this and
+            // that this type is not instantiated in any function of the module
+            char_type_candidate = Some((struct_name, struct_handle, def));
+            break; // no reason to look any further
+        }
+    }
+    for fn_def in &module.function_defs {
+        let fn_handle = module.function_handle_at(fn_def.function);
+        let fn_name = module.identifier_at(fn_handle.name);
+        if fn_name == INIT_FN_NAME {
+            if let Some((struct_name, struct_handle, _)) = char_type_candidate {
```

Comment:
nit, some naming here like `witness_name, witness_handle` would make this a bit more readable (at least to me)

I was very confused when looking at the other functions when I just saw `struct_name` and `struct_handle`

Reply:
Well, we are looking for structs that can potentially be one-time witness (characteristic) types but at this point it's not guaranteed that they are (they are still candidates subject to verification). Calling them `witness` here is a bit premature I thought...

Reply:
I think once you have the thing that is the `ascii_uppercase` of the module name, that is the characteristic/witness type. After that, you are ensuring it satisfies the necessary properties. 
But I hear you. At the very least, please some name other than `struct_name`, since that could be any struct! Not just your "candidate"